model/command_autocomplete.go

```diff
@@ -0,0 +1,415 @@
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"net/url"
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+// Argument types
+const (
+	AutocompleteArgTypeText        = "TextInput"
+	AutocompleteArgTypeStaticList  = "StaticList"
+	AutocompleteArgTypeDynamicList = "DynamicList"
+)
+
+// AutocompleteData describes slash command autocomplete information.
+type AutocompleteData struct {
+	// Trigger of the command
+	Trigger string
+	// Text displayed to the user to help with the autocomplete
+	HelpText string
+	// Role of the user who should be able to see the autocomplete info of this command
+	RoleID string
+	// Arguments of the command. Arguments can be named or positional.
+	// If they are positional order in the list matters, if they are named order does not matter.
+	// All arguments should be either named or positional, no mixing allowed.
+	Arguments []*AutocompleteArg
+	// Subcommands of the command
+	SubCommands []*AutocompleteData
+}
+
+// AutocompleteArg describes an argument of the command. Arguments can be named or positional.
+// If Name is empty string Argument is positional otherwise it is named argument.
+// Named arguments are passed as --Name Argument_Value.
+type AutocompleteArg struct {
+	// Name of the argument
+	Name string
+	// Text displayed to the user to help with the autocomplete
+	HelpText string
+	// Type of the argument
+	Type string
+	// Actual data of the argument (depends on the Type)
+	Data interface{}
+}
+
+// AutocompleteTextArg describes text user can input as an argument.
+type AutocompleteTextArg struct {
+	// Hint of the input text
+	Hint string
+	// Regex pattern to match
+	Pattern string
+}
+
+// AutocompleteStaticListItem describes an item in the AutocompleteStaticListArg.
+type AutocompleteStaticListItem struct {
+	Item     string
+	HelpText string
+}
+
+// AutocompleteStaticListArg is used to input one of the arguments from the list,
+// for example [yes, no], [on, off], and so on.
+type AutocompleteStaticListArg struct {
+	PossibleArguments []AutocompleteStaticListItem
+}
+
+// AutocompleteDynamicListArg is used when user wants to download possible argument list from the URL.
+type AutocompleteDynamicListArg struct {
+	FetchURL string
+}
+
+// AutocompleteSuggestion describes a single suggestion item sent to the frontend
+type AutocompleteSuggestion struct {
+	// Hint describes what user might want to input
+	Hint string
+	// Description of the command
+	Description string
+}
+
+// NewAutocompleteData returns new Autocomplete data.
+func NewAutocompleteData(trigger string, helpText string) *AutocompleteData {
+	return &AutocompleteData{
+		Trigger:     trigger,
+		HelpText:    helpText,
+		RoleID:      SYSTEM_USER_ROLE_ID,
+		Arguments:   []*AutocompleteArg{},
+		SubCommands: []*AutocompleteData{},
+	}
+}
+
+// AddCommand adds a subcommand to the autocomplete data.
+func (ad *AutocompleteData) AddCommand(command *AutocompleteData) {
```

Comment:
What happend if command is `nil`?

Reply:
Shouldn't this method be called `AddSubCommand`?

Reply:
If command is `nil` there will be a `panic`. It's probably better to check here and `panic` with appropriate message.

`AddCommand` is better name I think, it's also consistent with `cobra`

Reply:
Similar proposal as in https://github.com/mattermost/mattermost-server/pull/13602#discussion_r412110670